python/pyiceberg/expressions/base.py

```diff
@@ -672,3 +683,55 @@ def visit_unbound_predicate(self, predicate) -> BooleanExpression:
 
     def visit_bound_predicate(self, predicate) -> BooleanExpression:
         raise TypeError(f"Found already bound predicate: {predicate}")
+
+
+class BoundBooleanExpressionVisitor(BooleanExpressionVisitor[T], ABC):
+    @abstractmethod
+    def visit_in(self, ref: BoundReference[T], literals: set[Literal[Any]]) -> T:
```

Comment:
If we update this to Term then we won't need to implement `handle_non_reference` because visitors will be built to handle non-references.

Reply:
Rather than `ref` I think this should accept `term: BoundTerm[T]`